python/ray/data/_internal/execution/operators/output_splitter.py

```diff
@@ -0,0 +1,228 @@
+            object_store_memory=sum(b.size_bytes() for b in self._buffer)
+            + sum(b.size_bytes() for b in self._output_queue)
+        )
+
+    def progress_str(self) -> str:
+        if self._equal:
+            return f"{len(self._buffer)} buffered"
+        assert not self._buffer
+        return ""
+
+    def _dispatch_bundles(self) -> None:
+        # Dispatch all dispatchable bundles from the internal buffer.
+        # This may not dispatch all bundles when equal=True.
+        while self._buffer:
+            target_index = self._select_output_index()
+            target_bundle = self._pop_bundle_to_dispatch(target_index)
+            if self._can_safely_dispatch(target_index, target_bundle.num_rows()):
+                target_bundle.output_split_idx = target_index
+                self._num_output[target_index] += target_bundle.num_rows()
+                self._output_queue.append(target_bundle)
+            else:
+                # Put it back and abort.
+                self._buffer.insert(0, target_bundle)
+                break
+
+    def _select_output_index(self) -> int:
+        # Greedily dispatch to the consumer with the least data so far.
+        i, _ = min(enumerate(self._num_output), key=lambda t: t[1])
+        return i
+
+    def _pop_bundle_to_dispatch(self, target_index: int) -> RefBundle:
+        # TODO implement locality aware bundle selection.
+        return self._buffer.pop(0)
+
+    def _can_safely_dispatch(self, target_index: int, nrow: int) -> bool:
+        if not self._equal:
+            # If not in equals mode, dispatch away with no buffer requirements.
+            return True
+        output_distribution = self._num_output.copy()
+        output_distribution[target_index] += nrow
+        buffer_requirement = self._calculate_buffer_requirement(output_distribution)
+        buffer_size = sum(b.num_rows() for b in self._buffer)
+        return buffer_size >= buffer_requirement
+
+    def _calculate_buffer_requirement(self, output_distribution: List[int]) -> int:
+        # Calculate the new number of rows that we'd need to equalize the row
+        # distribution after the bundle dispatch.
+        max_n = max(output_distribution)
+        return sum([max_n - n for n in output_distribution])
+
+    def _split_from_buffer(self, nrow: int) -> List[RefBundle]:
+        output = []
+        acc = 0
+        while acc < nrow:
+            b = self._buffer.pop()
+            if acc + b.num_rows() <= nrow:
+                output.append(b)
+                acc += b.num_rows()
+            else:
+                left, right = _split(b, nrow - acc)
+                output.append(left)
+                acc += left.num_rows()
+                self._buffer.append(right)
+                assert acc == nrow, (acc, nrow)
+
+        assert sum(b.num_rows() for b in output) == nrow, (acc, nrow)
+        return output
+
+
+def _split(bundle: RefBundle, left_size: int) -> (RefBundle, RefBundle):
```

Comment:
Do we want to keep the out-of-plan/executor tasks? It seems difficult to not have this though.

Reply:
We could in principle make the executor aware of these, but they only happen at the very end, so I'm not sure it's worth the complexity. We could add it later if it's problematic.